type-check.ts

```diff
@@ -63,19 +73,81 @@ export type TypeError = {
   message: string
 }
 
-export function equalType(t1: Type, t2: Type) {
+// combine the elements of two arrays into an array of tuples.
+// DANGER: behaves weirdly with arrays of different length.
+function zip<A, B>(l1: Array<A>, l2: Array<B>) : Array<[A, B]> {
+  return l1.map((el, i) => [el, l2[i]]); 
+}
+
+export function equalType(env: GlobalTypeEnv, t1: Type, t2: Type) {
+  // If either of t1 or t2 are type-variables, then the actual type
+  // that they are currently instantiated to needs to be fetched from
+  // the environment before comparing the types.
```

Comment:
Yeah this seems pretty gnarly! I still wonder about monomorphizing _before_ doing this to just leverage the existing type checker. What benefit do we get from adding this whole new layer of env? Like what's a program we can handle with this approach but not by monomorphize-then-typecheck?

Or, what would be complicated in the implementation elsewhere if we didn't follow this order?

Reply:
We definitely contemplated between the two approaches -
1. Monomorphize everything before type-checking and then type-check the newly created monomorphized classes.
2. Type-check without monomorphizing and then monomorphize.

Ultimately the decision came down to a Language Design thing of how we wanted Generics to work ( which may not necessarily be the way python3 checkers do things ). Some of our thought process and understanding was influenced by https://www.tangramvision.com/blog/c-rust-generics-and-specialization. Basically our high-level conclusion was that the C++ way would to do things is 1 and the Rust way to do things is 2.

How this translates to our design choices is:
- We want a Generic Class to be type-checked and valid for all types satisfied by its constraints regardless of whether it is actually instantiated / what types it is actually instantiated with over the course of the program.

Example 1:

```python
T = TypeVar('T') # Unconstrained type-variable

class Foo(Generic[T]):
	x : T = __EMPTY__
	
	def foo(self: Foo[T], other: Foo[T]) -> T:
		return self.x + other.x
```

We want the above program to fail compilation since T is an unconstrained type variable and we don't know that the "+" operation is valid for it. Even though no objects of Foo are actually ever created. If we monomorphized before type-checking (currently we rely on variable definition's type annotations to monomorphize), there is nothing to monomorphize since we don't know any concrete types to actually instantiate T with. So Foo would disappear before type checking and we wouldn't get an error.

Example 2:

```python
T = TypeVar('T') # Unconstrained type-variable

class Foo(Generic[T]):
	x : T = __EMPTY__
	
	def foo(self: Foo[T], other: Foo[T]) -> T:
		return self.x + other.x
		
f : Foo[int] = None
```

We want this to fail for the same reasons as the previous example. Even though the case actually used in the program - 'int' is valid, we want Foo to type-check for all possible assignments/instantiations. Again if we monomorphized before type-checking, we would create only a Foo$int which would type-check and we would silently ignore the error. From a programmers point of view we were of the opinion that it is better to be strict and do this checking early rather than an error suddenly being thrown in the future when Foo[T] is instantiated for some other type that does not have the '+' operations.

- We want our Generic Classes to be "truly" monomorphized to the types they are instantiated to.

Example 3:

```python
T = TypeVar('T', bound=Superclass)

class Superclass(object):
	pass
	
class Subclass1(Superclass):
	pass
	
class Subclass2(Superclass):
	pass
	
class Foo(Generic[T]):
	x : T = __EMPTY__
	
f1: Foo[Subclass1] = None
f2: Foo[Subclass2] = None
```

This example uses the "upper bound" constraint that we hope to introduce as part of our Milestone2. In this case, we want Foo to type-check for whenever T is instantiated to any subclass of Superclass, but at the same time we want to eventual monomorphized versions of the class used to be Foo$Subclass1 and Foo$Subclass2 and not Foo$Superclass. If we monomorphized for Subclass1 and Subclass2 and then type-checked, we would potentially again fail to report an error if the code for Foo used some field/method that was present in Subclass1 and Subclass2 but not in Superclass.

Given these requirements/choices - the way we did it was simply the first that we landed on that seemed to work. Hopefully this clears things up a bit and justifies the additional complexity somewhat. Please let us know if you would still like to see any changes to reduce the amount of type-checker changes.